Add :test:function, :test:select, and :test:format function definitions (#769)
unicode-org/message-format-wg: test/README.md

- [Data model stringify tests](https://github.com/messageformat/messageformat/blob/11c95dab2b25db8454e49ff4daadb817e1d5b770/packages/mf2-messageformat/src/data-model/stringify.test.ts)
- [Formatting tests](https://github.com/messageformat/messageformat/blob/11c95dab2b25db8454e49ff4daadb817e1d5b770/packages/mf2-messageformat/src/messageformat.test.ts)

A [JSON schema](./schemas/) is included for the test files in this repository.

> For users of Visual Studio Code, a [settings file](./.vscode/settings.json) is included that enables schema validation while editing the test files.

Reviewer: Line breaks

Author: This was added by #767, after which I merged `main` to this branch.

A [JSON schema](./schemas/) is included for the test files in this repository.

For users of Visual Studio Code, a [settings file](./.vscode/settings.json) is included that enables schema validation while editing the test files.

## Test Functions

As the behaviour of some of the default registry _functions_
such as `:number` and `:datetime`
is dependent on locale-specific data and may vary between implementations,
the following _functions_ are defined for **test use only**:

### `:test:function`

This function is valid both as a _selector_ and as a _formatter_.

#### Operands

The function `:test:function` requires a [Number Operand](/spec/registry.md#number-operands) as its _operand_.

#### Options

> The only _option_ `:test:function` recognizes is `fd`,

Author: It started life as `fractionDigits`, but got shortened to `fd` corresponding with @aphillips requests to make this less like `:number`.

> The only _option_ `:test:function` recognizes is `fd`,

Reviewer 2: suggestion: rename `fd` to `ds`, and `FD` to `DS`,  so that it at least corresponds to "digit size"? 

@aphillips What say you?

> The only _option_ `:test:function` recognizes is `fd`,

Reviewer: My concern with `fractionDigits` was to stay away from confusing users with slightly different names (the ones in `:number` are `maximumFractionDigits` and `minimumFractionDigits`). Obscure names don't help anyone. I'd suggest `decimalPlaces`, or perhaps `decimalSize`. Alternatively, we could test drive number skeletons here.

> The only _option_ `:test:function` recognizes is `fd`,

Reviewer 2: `decimalPlaces` sgtm

> The only _option_ `:test:function` recognizes is `fd`,

Author: Done.

> a _digit size option_ for which only `0` and `1` are valid values.

Author: https://github.com/unicode-org/message-format-wg/blob/main/spec/registry.md#digit-size-options

> a _digit size option_ for which only `0` and `1` are valid values.

Reviewer 2: Linkifying digit size just as "Number Operand" is being linkified 5 lines above. Helps short circuit questions like mine for the reader.

```suggestion
a [_digit size option_](/spec/registry.md#digit-size-options) for which only `0` and `1` are valid values.
```

Reviewer: Underscore italics *are* linkified when the spec is turned into HTML. That's what [this](https://github.com/unicode-org/message-format-wg/blob/main/docs/tools/linkify.js) is for.

- If the `Input` is any other value, the method will return an empty list.

When an _expression_ with a `:test:function` _annotation_ is assigned to a _variable_ by a _declaration_
and that _variable_ is used as an _option_ value,
its resolved value is the `Input` value.

When `:test:function` is used as a _formatter_,
a _placeholder_ resolving to a value with a `:test:function` _expression_
is formatted as a concatenation of the following parts:

1. If `Input` is less than 0, the character `-` U+002D Hyphen-Minus.
1. The truncated absolute integer value of `Input`, i.e. floor(abs(`Input`)),
   formatted as a sequence of decimal digit characters (U+0030...U+0039).
1. If `FD` is 1, then
   1. The character `.` U+002E Full Stop.
   1. The single decimal digit character representing the value floor((abs(`Input`) - floor(abs(`Input`))) \* 10)

If the formatting target is a sequence of parts,
each of the above parts will be emitted separately
rather than being concatenated into a single string.

### `:test:select`

This _function_ accepts the same _operands_ and _options_,
and behaves exactly the same as `:test:function`,

> except that it cannot be used for formatting.

Reviewer 2: I think having a selection function that isn't capable of formatting is strange because formatting can be a precursor to selection, and that is indeed the case for plurals/numbers, and this test function takes numbers.

What is the motivation for having such a function for selection but not formatting?

One concern is that such a function implies that selection and formatting aren't linked. Another concern is that if the constraint that `:test:select` is defined by is not super useful, and if its functionality is a subset of the full `:test:function`, then we can remove `:test:select` altogether without loss of functionality.

Reviewer 2: Ok, that makes sense. IIUC, the `:test:select` function isn't so much about number formatting as much as standing in for the category of functions that fit the bill of select-only, for which there exist regular examples like gender. Sounds good.

When `:test:select` is used as a _formatter_,
a "not-formattable" error is emitted and the _placeholder_ is formatted with
a _fallback value_.

> ### `:test:format`

Reviewer 2: Similar comment as for `:test:select`: can we remove this as redundant given the superset functionality in `:test:function`?

Author: The intent is to allow testing the capability in MF2 of defining a function that's only a formatter, such as `:datetime` in the default registry. With `:test:format`, it becomes possible to validate the implementation's behaviour when such a function is used as a selector. That's not possible to test with `:test:function`.

Reviewer 2: Ok, that makes sense. Sounds good.

      or their corresponding string representations `'0'` or `'1'`, then
      1. Set _FD_ to be the numerical value of the option.
   1. Else if its value is not an unresolved value set by [Option Resolution](/spec/formatting.md#option-resolution),
      1. Emit "bad-option" Resolution Error.
      1. Use a fallback representation as the resolved value of the expression.

When `:test:function` is used as a selector,
the behaviour of calling it as the `rv` value of MatchSelectorKeys(`rv`, `keys`)
(see [Resolve Preferences](/spec/formatting.md#resolve-preferences) for more information)
depends on its _Input_ and _FD_ values.

- If the _Input_ is `1` and _FD_ is `1`,
  the method will return some slice of the list « `'1.0'`, `'1'` »,
  depending on whether those values are included in `keys`.
- If the _Input_ is `1` and _FD_ is `0`,
  the method will return the list « `'1'` » if `keys` includes `'1'`, or an empty list otherwise.
- If the _Input_ is any other value, the method will return an empty list.

When an expression with a `:test:function` annotation is assigned to a variable by a declaration
and that variable is used as an option value,

> its resolved value is the _Input_ value.

Reviewer: ...line breaks and italics for terms...

Author: Introducing italicized terms here would make it more difficult to discern the local _Input_ reference, which I'd prefer to avoid. Can it be avoided, or is there some other styling I should Use for _Input_ and _FD_?

Reviewer: Use `code` like `Input`, perhaps??

When `:test:function` is used as a formatter,
a placeholder resolving to a value with a `:test:function` expression
is formatted as a concatenation of the following parts:

1. If _Input_ is less than `0`, the character `-` U+002D Hyphen-Minus.
1. The truncated absolute integer value of _Input_, i.e. floor(abs(_Input_)),

>    formatted as a sequence of decimal digit caharacters.

Reviewer: Typo

```suggestion
   formatted as a sequence of decimal digit characters (U+0030...U+0039).
```

1. If _FD_ is `1`, then
   1. The character `.` U+002E Full Stop.
   1. The single decimal digit character representing the value floor((abs(_Input_) - floor(abs(_Input_))) \* 10)

If the formatting target is a sequence of parts,
each of the above parts will be emitted separately
rather than being concatenated into a single string.

> Note that for purposes of clarity, the formatting of `:test:function` does not perform any rounding.

Reviewer: This shouldn't be a note as we are describing the implementation.

```suggestion
`:test:function` does not round the formatted value.
Implementations MAY impose an upper limit on the number of digits rendered,
based on local requirements (such as the size of the underlying decimal type)
```

Author: The proposed MAY is incorrect, as the formatted digits are exactly defined above. If this is unclear, let's just remove it?
```suggestion
```

- [CST parse/stringify tests](https://github.com/messageformat/messageformat/blob/11c95dab2b25db8454e49ff4daadb817e1d5b770/packages/mf2-messageformat/src/cst/cst.test.ts)
- [Data model stringify tests](https://github.com/messageformat/messageformat/blob/11c95dab2b25db8454e49ff4daadb817e1d5b770/packages/mf2-messageformat/src/data-model/stringify.test.ts)
- [Formatting tests](https://github.com/messageformat/messageformat/blob/11c95dab2b25db8454e49ff4daadb817e1d5b770/packages/mf2-messageformat/src/messageformat.test.ts)

## Test Functions

As the behaviour of `:number`, `:datetime`, and other default message functions
is dependent on locale-specific data and may vary between implementations,
the following functions are defined for test use only:

### `test:number`

This function is valid both as a selector and as a formatter.
The only option it recognizes is `fractionDigits`,

> all other option values passed to it are ignored.

Reviewer: How would I use this?

A test like this still doesn't work:

```
You have {$num :number} elderberries.
```

I have to rewrite it to use `:test:number` and `:test:number` has different options from `:number`. Implementations cannot test full conformance with MF2 using the tests because default registry is included in full conformance.

Maybe a better approach would be to have an invariant locale or a test locale and to define what the outputs for that locale are. e.g. we could use the root locale (`und`) or possibly a patch on that (`und-x-test`) [note that pseudo-translation locales recommend themselves here as well--especially the potential for RTL pseudo-locales so both LTR and RTL messages can be tested]

Possible we should consider doing design first?

- [CST parse/stringify tests](https://github.com/messageformat/messageformat/blob/11c95dab2b25db8454e49ff4daadb817e1d5b770/packages/mf2-messageformat/src/cst/cst.test.ts)
- [Data model stringify tests](https://github.com/messageformat/messageformat/blob/11c95dab2b25db8454e49ff4daadb817e1d5b770/packages/mf2-messageformat/src/data-model/stringify.test.ts)
- [Formatting tests](https://github.com/messageformat/messageformat/blob/11c95dab2b25db8454e49ff4daadb817e1d5b770/packages/mf2-messageformat/src/messageformat.test.ts)

## Test Functions

> As the behaviour of `:number`, `:datetime`, and other default message functions

Reviewer: There's a specific list. I think `:string` shouldn't be on that list?

```suggestion
As the behavior of the functions 
`:number`,
`:integer`,
`:datetime`,
`:date`,
and `:time`
in the default registry depend 
on specific versions of locale data and on their local implementation
the following functions are defined for **test use only**.
```

Author: Rewrote it a bit, still leaving the list open so that adding a new default registry function doesn't require updating this as well.

is dependent on locale-specific data and may vary between implementations,
the following functions are defined for **test use only**:

> ### `test:func`

Reviewer: `select` and `format` are fine, but `func` should probably be `function`

```suggestion
### `:test:function`
```

This function is valid both as a selector and as a formatter.

> The only option it recognizes is `fd`,

Reviewer: What does `fd` mean?

We should probably follow the model in the registry.md document in defining these functions. There we have subsections

- operands
- options
- selection
- formatting
- (other stuff)

I think a little bit of design is probably useful here as well. It would be nice to be able to have `:function` (and `:select`) model `:number`'s behavior (with both exact and keyword matching) by defining what numbers match the plural keywords `zero`, `one`, `two`, `few`, `many`, and `other`. That way we can provide tests that implementers can use to see if they match the sorting mechanism in `formatting.md` without writing number/plural selection over.

Another thing to consider is that we don't need options if we can specify how `:function` and `:select` work internally. One thing we could specify is that they'll receive an ordered list or other set of arguments from the test (not in the message) that can be used to program their behavior in a simplistic way, e.g. a regexp that should match the operand.

Author: > What does `fd` mean?

It started life as `fractionDigits`, but you asked for this to be distanced from the real `:number` implementation, and I simplified its values to just `0` and `1`, so it makes sense to be a bit more abstract. So now it's just an option that has _some_ effect on the `:test:function` behaviour, and a small set of valid values.

> We should probably follow the model in the registry.md document in defining these functions. There we have subsections

That model doesn't exactly work here, because here we need to be explicit about exactly what this function does, and how it behaves in different settings.

> It would be nice to be able to have `:function` (and `:select`) model `:number`'s behavior (with both exact and keyword matching) by defining what numbers match the plural keywords `zero`, `one`, `two`, `few`, `many`, and `other`. That way we can provide tests that implementers can use to see if they match the sorting mechanism in `formatting.md` without writing number/plural selection over.

That's already supported. Selection on `{1 :test:function}` will only match the `1` key, while selection on `{1 :test:function fd=1}` will match both `1` and `1.0` keys, preferring the latter. This is explicitly different from `:number` selection, again, because you asked for this to be different.

> Another thing to consider is that we don't need options if we can specify how `:function` and `:select` work internally. One thing we could specify is that they'll receive an ordered list or other set of arguments from the test (not in the message) that can be used to program their behavior in a simplistic way, e.g. a regexp that should match the operand.

I would prefer working iteratively, and first seeing where this small family of functions gets us. If/as we discover parts of the spec that do need testing but can't be tested with these functions, then we can reconsider.

I would very much prefer not needing to specify even simple programmable behaviour at the level needed to guarantee exact behaviour across all implementations. For this, regexp is pretty much right out unless there's a ready reference somewhere for a minimum subset that's supported everywhere _and_ for which a validator is available to keep the test data honest.

Reviewer: > but you asked for this to be distanced from the real :number implementation

Right. My thinking was that we want something that can be used for conformance testing now and later. We want it to be easy to implement, have deterministic behavior, and cover selection/formatting features required in our spec.

We have [an example](https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md#example-3) in our spec now that should be testable. The example uses `:number`, but replacing it with `:function` could be straightforward.

> That model doesn't exactly work here, because here we need to be explicit about exactly what this function does, and how it behaves in different settings.

That model can work here, because we can describe the operands and options and the proceed to describe specifically how selection works and how formatting works. You're correct that, unlike `:number` or `:datetime`, we need to provide very specific implementation instructions.

Bear in mind that we need to flesh out how to write proposals for the default and "RGI" registry. The test functions should probably go through the motions and use the forms of that process.

> I would very much prefer not needing to specify even simple programmable behaviour at the level needed to guarantee exact behaviour across all implementations. For this, regexp is pretty much right out unless there's a ready reference somewhere for a minimum subset that's supported everywhere and for which a validator is available to keep the test data honest.

Fair enough.

Author: > We have [an example](https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md#example-3) in our spec now that should be testable. The example uses `:number`, but replacing it with `:function` could be straightforward.

Sure, that should be pretty directly portable to using `:test:function`. I'll be happy to do so after #767 lands, either within this PR or in a follow-up.

> Bear in mind that we need to flesh out how to write proposals for the default and "RGI" registry. The test functions should probably go through the motions and use the forms of that process.

So could we land this first with its current freeform description, and then update it once the forms of that process are indeed fleshed out? I would really prefer not blocking improvements in tests on figuring out a new proposal process.

Reviewer: > So could we land this first with its current freeform description, and then update it once the forms of that process are indeed fleshed out? I would really prefer not blocking improvements in tests on figuring out a new proposal process.

Sure, we can land this. 

However, nothing we do here will effect v45. We have _lots_ to do in the next few months. Each thing we do should not require later additional work to the degree reasonable.

Note that I'm not asking you to wait on a proposal process, but to try and use the flabby outlines of what we have to see what we can learn towards creating said process. If it's too hard, we probably need to look in another direction. FWIW, it would be amusing to see the registry XML or JSON description for `:function`, `:select`, and `:format`

Author: > Note that I'm not asking you to wait on a proposal process, but to try and use the flabby outlines of what we have to see what we can learn towards creating said process.

Done.

all other option values passed to it are ignored.

When resolving a `test:func` expression,
its _Input_ and _FD_ values are determined as follows:

1. Let _FD_ be `0`.

> 1. Let _arg_ be the resolved value of the expression argument.

Reviewer: "expression argument"? Is that an "operand"?
